### PR TITLE
Graphiql enhancements

### PR DESCRIPTION
**Pull request by younky-yang**
## Hey, I just made a Pull Request!

This PR tries to enhance the graphiql api and API-doc for GraphQL api. It enables user to try the graphql api in API definition page if they specify the annotation "backstage.io/api-graphql-url" in entity's definition. It also populates all these graphQL API on Graphiql plugin, so user can easily to play with these APIs. 

https://github.com/backstage/backstage/issues/19370

#### :heavy_check_mark: Checklist

- [*] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [*] Tests for new functionality and regression tests for bug fixes
- [*] Screenshots attached (for UI changes)
- [*] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
image
image

### CODE REVIEW

File: .changeset/heavy-toys-pull.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+---
+'@backstage/catalog-model': minor
```
Comment (freben) [inline .changeset/heavy-toys-pull.md:2]: the examples aren't shipped with the package - remove this line

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx [hunk 2]
```diff
@@ -46,17 +53,46 @@ type Props = {
 
 export const GraphQlDefinition = ({ definition }: Props) => {
   const classes = useStyles();
+  const { entity } = useEntity();
```
Comment (freben) [inline plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx:56]: This is a major change in contract for the component. It used to be possible to render it outside of an entity page, and now you can't. If people were using this component standalone to render a graphql definition, their Backstage will suddenly break after upgrading.

I suggest adding the endpoint URL to its props instead, as an optional string. Then leave it up to the surroundings to supply if any way they see fit, if they want to

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx [hunk 3]
```diff
@@ -46,17 +53,46 @@ type Props = {
...
   const schema = buildSchema(definition);
 
+  const url = entity.metadata.annotations?.['backstage.io/api-graphql-url'];
```
Comment (freben) [inline plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx:59]: maybe `endpoint` instead of `url`? 🤔 matter of taste i guess

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx [hunk 4]
```diff
@@ -46,17 +53,46 @@ type Props = {
...
+          fetcher={
+            url
+              ? async (params: any, options: any = {}) => {
+                  const body = JSON.stringify(params);
+                  const headers = {
+                    'Content-Type': 'application/json',
+                    ...options.headers,
+                  };
+
+                  const res = await fetch(url!, {
```
Comment (freben) [inline plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx:74]: frontend components should prefer to use the fetch that comes out of `useApi(fetchApiRef)` instead of the native one

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx [hunk 5]
```diff
@@ -46,17 +53,46 @@ type Props = {
...
+                    body,
+                  });
+                  return res.json();
+                }
+              : () => Promise.resolve(null) as any
+          }
           schema={schema}
           docExplorerOpen
           defaultSecondaryEditorOpen={false}
+          headerEditorEnabled
```
Comment (freben) [inline plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx:86]: is this an unrelated fix?

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx [hunk 6]
```diff
@@ -38,6 +39,12 @@ const useStyles = makeStyles<BackstageTheme>(() => ({
       },
     },
   },
+  hints: {
```
Comment (freben) [inline plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx:42]: hmm you probably want to use `<Alert` from material-ui at info level instead of hand-winging this, since that gives consistent styling across the app

File: yarn.lock [hunk 7]
```diff
@@ -27412,16 +27413,16 @@ __metadata:
   linkType: hard
 
 "graphql-language-service@npm:^5.0.6":
-  version: 5.0.6
-  resolution: "graphql-language-service@npm:5.0.6"
+  version: 5.1.7
+  resolution: "graphql-language-service@npm:5.1.7"
```
Comment (freben) [inline yarn.lock:27417]: A lot of upgrades in here seem unrelated. Maybe revert yarn.lock to its master shape and then run yarn install again?

File: plugins/graphiql/package.json [hunk 8]
```diff
@@ -34,12 +34,13 @@
...
     "@backstage/core-components": "workspace:^",
     "@backstage/core-plugin-api": "workspace:^",
+    "@backstage/plugin-catalog-react": "workspace:^",
     "@backstage/theme": "workspace:^",
     "@material-ui/core": "^4.12.2",
     "@material-ui/icons": "^4.9.1",
     "@material-ui/lab": "4.0.0-alpha.61",
     "@types/react": "^16.13.1 || ^17.0.0",
-    "graphiql": "^1.5.12",
+    "graphiql": "^1.8.8",
```
Comment (freben) [inline plugins/graphiql/package.json:43]: was this needed? just checking, since mixing concerns adds to the risk of a revert of it all if something goes wrong

File: .changeset/heavy-toys-pull.md [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+'@backstage/plugin-api-docs': minor
+'@backstage/plugin-graphiql': minor
+---
+
+Enhanced API-docs for graphql to enable user try out the api if they provide backstage.io/api-graphql-url annotation in catalog entity.
```
Comment (freben) [inline .changeset/heavy-toys-pull.md:7]:
```suggestion
Enhanced the `api-docs` plugin for graphql to enable users to try out the api if they provide a `backstage.io/api-graphql-url` annotation in their catalog entity.
```

File: plugins/graphiql/src/lib/api/GraphQLEndpoints.ts [hunk 10]
```diff
@@ -76,6 +77,30 @@ export class GraphQLEndpoints implements GraphQLBrowseApi {
     };
   }
 
+  // Create the GraphQLEndpoints from the catalog entities with kind=api and type = graphql.
+  static async fromCatalogEntities(
```
Comment (freben) [inline plugins/graphiql/src/lib/api/GraphQLEndpoints.ts:81]: This should not be a public method. The API already has a `getEndpoints` which is the one that should return the results. The fact that they came from one source or another should be entirely hidden to the `GraphiQLPage`. Make it the responsibility of the class to initialize itself in any way it sees fit and hide how it deals with the enumeration of urls

File: plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx (hunk 5 above)
Comment (younky-yang) [reply plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinition.tsx:86]: this is to ensure the header editor always enabled, so we can pass some auth information through header.

File: plugins/graphiql/package.json (hunk 8 above)
Comment (younky-yang) [reply plugins/graphiql/package.json:43]: to keep this consistent of the one used in api-docs

**Comment by younky-yang** [comment]
@freben, thanks for reviewing this, will refine the PR to address your comments.

**Comment by benjdlambert** [comment]
@younky-yang I think this PR has gotten a little confused here. I'd recommend rebasing or open up a new PR entirely.

**Comment by younky-yang** [comment]
> @younky-yang I think this PR has gotten a little confused here. I'd recommend rebasing or open up a new PR entirely.

yes，just rebased with master and will address the comments and then push changes.
